File: sportsverse/src/services/api/cricket.ts
/**
 * Cricket Data API Service
 * Communicates with SportsVerse backend at http://localhost:8000
 * All data comes from Highlightly real cricket API
 */

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1'

export interface ApiResponse<T> {
  success: boolean
  data: T
  message: string
  timestamp: string
}

// ==================== REAL HIGHLIGHTLY API TYPES ====================

export interface BatterRecord {
  player: {
    id: string
    name: string
    battingStyles: string[]
    bowlingStyles: string[]
    roles: string[]
  }
  runs: number
  balls: number
  fours: number
  sixes: number
  dismissalStatus: string
  battingStrikeRate: number
  dismissalFielders: Array<{
    name: string
    isKeeper?: boolean
    isSubstitute?: boolean
  }>
}

export interface BowlerRecord {
  player: {
    name: string
    battingStyles: string[]
    bowlingStyles: string[]
    roles: string[]
  }
  overs: number
  wickets: number
  concededRuns: number
  economy: number
  maidens: number
}

export interface WicketRecord {
  order: number
  runs: number
  overs: number
  dismissalBatsman: {
    name: string
  }
}

export interface PartnershipRecord {
  runs: number
  balls: number
  overs: number
  firstPlayer: { id: string; name: string }
  secondPlayer: { id: string; name: string }
  firstPlayerRuns: number
  firstPlayerBalls: number
  secondPlayerRuns: number
  secondPlayerBalls: number
}

export interface InningStatistics {
  inningNumber: number
  team: {
    id: number
    name: string
    abbreviation: string
    logo: string
    byes: number
    legByes: number
    wides: number
    noBalls: number
    extras: number
    fours?: number | null
    sixes?: number | null
  }
  inningBatsmen: BatterRecord[]
  inningBowlers: BowlerRecord[]
  fallOfWickets: WicketRecord[]
  inningPartnerships: PartnershipRecord[]
  fieldingSummary?: {
    catches: number | null
    runOuts: number | null
    stumpings: number | null
    runsSaved: number | null
    catchesDropped: number | null
  }
}

export interface SquadInfo {
  team: {
    id: number
    name: string
    abbreviation: string
    logo: string
  }
  players: Array<{
    name: string
    battingStyles: string[]
    bowlingStyles: string[]
    roles: string[]
  }>
}

export interface BestPerformer {
  team: {
    id: number
    name: string
    abbreviation: string
    logo: string
  }
  players: Array<{
    name: string
    statistics: Record<string, unknown>
  }>
}

export interface HighlightlyMatch {
  id: string
  startDate: string
  startTime: string
  endDate: string
  format: string
  dayType: string
  country: {
    code: string
    name: string
    logo: string
  }
  homeTeam: {
    id: string | number
    name: string
    abbreviation: string
    logo: string
  }
  awayTeam: {
    id: string | number
    name: string
    abbreviation: string
    logo: string
  }
  state: {
    description: string
    report?: string
    teams: {
      home: {
        score: string
        info: string | null
      }
      away: {
        score: string
        info: string | null
      }
    }
  }
  league: {
    id: string | number
    name: string
    season?: number
    logo?: string | null
  }
  venue: {
    name: string | null
    city: string | null
    country: string | null
    capacity: string | null
  }
  forecast: {
    status: string | null
    temperature: string | null
  }
  predictions: unknown | null
  statistics: InningStatistics[]
  squad: SquadInfo[]
  bestBatsmen?: BestPerformer[]
  bestBowlers?: BestPerformer[]
}

// ==================== LEGACY SIMPLIFIED TYPES ====================

export interface Match {
  id: string
  external_id?: string
  homeTeam: {
    id?: string
    name: string
    shortName?: string
    logo?: string
  }
  awayTeam: {
    id?: string
    name: string
    shortName?: string
    logo?: string
  }
  status: 'live' | 'upcoming' | 'completed' | 'abandoned'
  format: string
  venue?: {
    id?: string
    name?: string
    city?: string
    country?: string
  }
  league?: {
    id?: string
    name: string
  }
  series?: {
    id?: string
    name?: string
  }
  matchDate?: string
  startTime?: string
  liveData?: {
    homeTeam?: {
      score?: string
      info?: string
    }
    awayTeam?: {
      score?: string
      info?: string
    }
    description?: string
    report?: string
  }
  scorecard?: unknown
  commentary?: unknown
}


// ==================== API FUNCTIONS ====================

export async function getLiveMatches(): Promise<Match[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/live`)
    if (!response.ok) throw new Error(`Failed to fetch live matches: ${response.status}`)
    const data: ApiResponse<Match[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching live matches:', error)
    return []
  }
}

export async function getTodayMatches(): Promise<Match[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/today`)
    if (!response.ok) throw new Error(`Failed to fetch today's matches: ${response.status}`)
    const data: ApiResponse<Match[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching today\'s matches:', error)
    return []
  }
}

export async function getUpcomingMatches(days: number = 7): Promise<Match[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/upcoming?days=${days}`)
    if (!response.ok) throw new Error(`Failed to fetch upcoming matches: ${response.status}`)
    const data: ApiResponse<Match[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching upcoming matches:', error)
    return []
  }
}

export async function getResults(limit: number = 20): Promise<Match[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/results?limit=${limit}`)
    if (!response.ok) throw new Error(`Failed to fetch results: ${response.status}`)
    const data: ApiResponse<Match[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching results:', error)
    return []
  }
}

export async function getMatchDetails(matchId: string): Promise<HighlightlyMatch | null> {
  try {
    const response = await fetch(`${API_BASE}/cricket/match/${matchId}`)
    if (!response.ok) {
      if (response.status === 404) return null
      throw new Error(`Failed to fetch match details: ${response.status}`)
    }
    const data: ApiResponse<HighlightlyMatch> = await response.json()
    return data.data || null
  } catch (error) {
    console.error('Error fetching match details:', error)
    return null
  }
}

export async function getSeries(seriesId: string): Promise<Match | null> {
  try {
    const response = await fetch(`${API_BASE}/cricket/series/${seriesId}`)
    if (!response.ok) {
      if (response.status === 404) return null
      throw new Error(`Failed to fetch series details: ${response.status}`)
    }
    const data: ApiResponse<Match> = await response.json()
    return data.data || null
  } catch (error) {
    console.error('Error fetching series:', error)
    return null
  }
}

export async function getSeriesMatches(seriesId: string, status: string = 'all'): Promise<Match[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/series/${seriesId}/matches?status=${status}`)
    if (!response.ok) throw new Error(`Failed to fetch series matches: ${response.status}`)
    const data: ApiResponse<Match[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching series matches:', error)
    return []
  }
}

export async function getStandings(seriesId: string): Promise<unknown[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/standings/${seriesId}`)
    if (!response.ok) throw new Error(`Failed to fetch standings: ${response.status}`)
    const data: ApiResponse<unknown[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching standings:', error)
    return []
  }
}

export async function getNews(limit: number = 20): Promise<unknown[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/news?limit=${limit}`)
    if (!response.ok) throw new Error(`Failed to fetch news: ${response.status}`)
    const data: ApiResponse<unknown[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching news:', error)
    return []
  }
}

export async function searchCricket(query: string): Promise<{
  matches: Match[]
  players: unknown[]
  teams: unknown[]
  news: unknown[]
}> {
  try {
    const response = await fetch(`${API_BASE}/cricket/search?q=${encodeURIComponent(query)}`)
    if (!response.ok) throw new Error(`Failed to search: ${response.status}`)
    const data: ApiResponse<{
      matches: Match[]
      players: unknown[]
      teams: unknown[]
      news: unknown[]
    }> = await response.json()
    return data.data || { matches: [], players: [], teams: [], news: [] }
  } catch (error) {
    console.error('Error searching cricket data:', error)
    return { matches: [], players: [], teams: [], news: [] }
  }
}

export async function getTeam(teamId: string): Promise<unknown> {
  try {
    const response = await fetch(`${API_BASE}/cricket/team/${teamId}`)
    if (!response.ok) {
      if (response.status === 404) return null
      throw new Error(`Failed to fetch team: ${response.status}`)
    }
    const data: ApiResponse<unknown> = await response.json()
    return data.data || null
  } catch (error) {
    console.error('Error fetching team:', error)
    return null
  }
}

export async function getTeamSquad(teamId: string): Promise<unknown[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/team/${teamId}/squad`)
    if (!response.ok) throw new Error(`Failed to fetch team squad: ${response.status}`)
    const data: ApiResponse<unknown[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching team squad:', error)
    return []
  }
}

export async function getTeamFixtures(teamId: string): Promise<Match[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/team/${teamId}/fixtures`)
    if (!response.ok) throw new Error(`Failed to fetch team fixtures: ${response.status}`)
    const data: ApiResponse<Match[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching team fixtures:', error)
    return []
  }
}

export async function getTeamResults(teamId: string): Promise<Match[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/team/${teamId}/results`)
    if (!response.ok) throw new Error(`Failed to fetch team results: ${response.status}`)
    const data: ApiResponse<Match[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching team results:', error)
    return []
  }
}

export async function getTeamNews(teamId: string): Promise<unknown[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/team/${teamId}/news`)
    if (!response.ok) throw new Error(`Failed to fetch team news: ${response.status}`)
    const data: ApiResponse<unknown[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching team news:', error)
    return []
  }
}

export async function getPlayer(playerId: string): Promise<unknown> {
  try {
    const response = await fetch(`${API_BASE}/cricket/player/${playerId}`)
    if (!response.ok) {
      if (response.status === 404) return null
      throw new Error(`Failed to fetch player: ${response.status}`)
    }
    const data: ApiResponse<unknown> = await response.json()
    return data.data || null
  } catch (error) {
    console.error('Error fetching player:', error)
    return null
  }
}

export async function getPlayerCareer(playerId: string): Promise<unknown> {
  try {
    const response = await fetch(`${API_BASE}/cricket/player/${playerId}/career`)
    if (!response.ok) throw new Error(`Failed to fetch player career: ${response.status}`)
    const data: ApiResponse<unknown> = await response.json()
    return data.data || null
  } catch (error) {
    console.error('Error fetching player career:', error)
    return null
  }
}

export async function getPlayerMatches(playerId: string): Promise<Match[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/player/${playerId}/matches`)
    if (!response.ok) throw new Error(`Failed to fetch player matches: ${response.status}`)
    const data: ApiResponse<Match[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching player matches:', error)
    return []
  }
}

export async function getPlayerStatistics(playerId: string): Promise<unknown> {
  try {
    const response = await fetch(`${API_BASE}/cricket/player/${playerId}/statistics`)
    if (!response.ok) throw new Error(`Failed to fetch player statistics: ${response.status}`)
    const data: ApiResponse<unknown> = await response.json()
    return data.data || null
  } catch (error) {
    console.error('Error fetching player statistics:', error)
    return null
  }
}

export async function getPlayerNews(playerId: string): Promise<unknown[]> {
  try {
    const response = await fetch(`${API_BASE}/cricket/player/${playerId}/news`)
    if (!response.ok) throw new Error(`Failed to fetch player news: ${response.status}`)
    const data: ApiResponse<unknown[]> = await response.json()
    return data.data || []
  } catch (error) {
    console.error('Error fetching player news:', error)
    return []
  }
}


// Alias for compatibility
export const getCricketNews = getNews

File: backend/app/providers/cricket.py
"""Cricket data provider with mock implementations"""
from typing import Optional, List
from datetime import datetime, timedelta
from app.providers.base import SportsProvider


class CricketProvider(SportsProvider):
    """
    Cricket provider implementation.
    Uses mock data for now. Will integrate real APIs later.
    """

    def __init__(self):
        self.name = "Cricket Provider"
        self.version = "1.0.0"

    async def get_live_matches(self) -> List[dict]:
        """Get all live matches - mock implementation."""
        return [
            {
                "id": "match_001",
                "homeTeam": {
                    "id": "team_1",
                    "name": "England",
                    "shortName": "ENG",
                    "country": "England",
                    "logo": "https://via.placeholder.com/64",
                },
                "awayTeam": {
                    "id": "team_2",
                    "name": "India",
                    "shortName": "IND",
                    "country": "India",
                    "logo": "https://via.placeholder.com/64",
                },
                "status": "live",
                "format": "T20I",
                "venue": {
                    "id": "venue_1",
                    "name": "The Oval",
                    "city": "London",
                    "country": "England",
                },
                "series": {
                    "id": "series_1",
                    "name": "England Tour of India",
                    "slug": "england-tour-india-2026",
                },
                "matchDate": datetime.now().isoformat(),
                "startTime": datetime.now().isoformat(),
                "liveData": {
                    "batting": {
                        "runs": 152,
                        "wickets": 4,
                        "overs": 15,
                        "balls": 2,
                        "runRate": 10.1,
                    },
                    "bowling": {
                        "wickets": 4,
                        "runs": 152,
                    },
                    "partnership": {
                        "runs": 35,
                        "balls": 28,
                    },
                    "target": 180,
                    "requiredRunRate": 9.4,
                    "winProbability": {"team1": 42, "team2": 58},
                },
                "tossInfo": {
                    "winner": "India",
                    "decision": "bat",
                    "message": "India won the toss and chose to bat",
                },
            }
        ]

    async def get_today_matches(self) -> List[dict]:
        """Get all matches happening today - mock implementation."""
        return [
            {
                "id": "match_002",
                "homeTeam": {"id": "team_3", "name": "Pakistan", "shortName": "PAK"},
                "awayTeam": {"id": "team_4", "name": "Bangladesh", "shortName": "BAN"},
                "status": "upcoming",
                "format": "T20I",
                "matchDate": datetime.now().isoformat(),
                "startTime": (datetime.now() + timedelta(hours=3)).isoformat(),
            },
            {
                "id": "match_003",
                "homeTeam": {"id": "team_5", "name": "South Africa", "shortName": "SA"},
                "awayTeam": {"id": "team_6", "name": "West Indies", "shortName": "WI"},
                "status": "upcoming",
                "format": "ODI",
                "matchDate": datetime.now().isoformat(),
                "startTime": (datetime.now() + timedelta(hours=8)).isoformat(),
            },
        ]

    async def get_upcoming_matches(self, days: int = 7) -> List[dict]:
        """Get upcoming matches for next N days - mock implementation."""
        matches = []
        for i in range(min(days * 2, 14)):
            match_date = datetime.now() + timedelta(days=i // 2)
            matches.append(
                {
                    "id": f"match_upcoming_{i}",
                    "homeTeam": {
                        "id": f"team_{i}",
                        "name": f"Team A {i}",
                        "shortName": f"TA{i}",
                    },
                    "awayTeam": {
                        "id": f"team_{i+10}",
                        "name": f"Team B {i}",
                        "shortName": f"TB{i}",
                    },
                    "status": "upcoming",
                    "format": ["T20I", "ODI", "Test"][i % 3],
                    "matchDate": match_date.isoformat(),
                    "startTime": match_date.isoformat(),
                }
            )
        return matches

    async def get_results(self, limit: int = 20) -> List[dict]:
        """Get recent completed matches - mock implementation."""
        matches = []
        for i in range(min(limit, 20)):
            match_date = datetime.now() - timedelta(days=i + 1)
            matches.append(
                {
                    "id": f"match_result_{i}",
                    "homeTeam": {
                        "id": f"team_{i}",
                        "name": f"Team A {i}",
                        "shortName": f"TA{i}",
                    },
                    "awayTeam": {
                        "id": f"team_{i+10}",
                        "name": f"Team B {i}",
                        "shortName": f"TB{i}",
                    },
                    "status": "completed",
                    "format": ["T20I", "ODI", "Test"][i % 3],
                    "matchDate": match_date.isoformat(),
                    "startTime": match_date.isoformat(),
                    "result": {
                        "winnerTeamId": f"team_{i}",
                        "resultType": "wickets",
                        "margin": 5 + i,
                    },
                }
            )
        return matches

    async def get_match(self, match_id: str) -> Optional[dict]:
        """Get detailed information about a specific match - mock implementation."""
        return {
            "id": match_id,
            "homeTeam": {
                "id": "team_1",
                "name": "England",
                "shortName": "ENG",
                "country": "England",
                "logo": "https://via.placeholder.com/64",
            },
            "awayTeam": {
                "id": "team_2",
                "name": "India",
                "shortName": "IND",
                "country": "India",
                "logo": "https://via.placeholder.com/64",
            },
            "status": "live",
            "format": "T20I",
            "venue": {
                "id": "venue_1",
                "name": "The Oval",
                "city": "London",
                "country": "England",
                "capacity": 29000,
            },
            "series": {
                "id": "series_1",
                "name": "England Tour of India",
                "slug": "england-tour-india-2026",
                "format": "T20",
                "status": "live",
            },
            "matchDate": datetime.now().isoformat(),
            "startTime": datetime.now().isoformat(),
            "liveData": {
                "innings": [
                    {
                        "team": "India",
                        "runs": 152,
                        "wickets": 4,
                        "overs": 15.2,
                    }
                ],
                "currentBatsmen": [
                    {"name": "Player A", "runs": 45, "balls": 38},
                    {"name": "Player B", "runs": 32, "balls": 25},
                ],
                "currentBowler": {"name": "Bowler A", "wickets": 2, "runs": 28},
                "partnership": {"runs": 35, "balls": 28},
                "target": 180,
                "requiredRunRate": 9.4,
                "winProbability": {"team1": 42, "team2": 58},
                "lastOver": [1, 2, 0, 4, 1, 1],
            },
            "tossInfo": {
                "winner": "India",
                "decision": "bat",
                "message": "India won the toss and chose to bat",
            },
            "weatherInfo": {
                "condition": "Partly cloudy",
                "temperature": 18,
                "humidity": 65,
                "windSpeed": 12,
            },
        }

    async def get_match_commentary(self, match_id: str) -> List[dict]:
        """Get commentary for a match - mock implementation."""
        return [
            {
                "id": f"com_{i}",
                "over": 15,
                "ball": 2,
                "batter": "Player A",
                "bowler": "Bowler A",
                "runs": 1,
                "text": "Single to mid-off",
                "timestamp": datetime.now().isoformat(),
            }
            for i in range(10)
        ]

    async def get_match_scorecard(self, match_id: str) -> Optional[dict]:
        """Get scorecard for a match - mock implementation."""
        return {
            "id": match_id,
            "innings": [
                {
                    "team": "India",
                    "runs": 152,
                    "wickets": 4,
                    "overs": 15.2,
                    "batsmen": [
                        {"name": "Player A", "runs": 45, "balls": 38, "fours": 3, "sixes": 1},
                        {"name": "Player B", "runs": 32, "balls": 25, "fours": 2, "sixes": 1},
                    ],
                }
            ],
        }

    async def get_player(self, player_id: str) -> Optional[dict]:
        """Get player details and statistics - mock implementation."""
        return {
            "id": player_id,
            "name": "Virat Kohli",
            "shortName": "Kohli",
            "country": "India",
            "role": "batsman",
            "jerseyNumber": 18,
            "photoUrl": "https://via.placeholder.com/200",
            "stats": {
                "t20i": {"runs": 3000, "avg": 40.5, "matches": 80},
                "odi": {"runs": 13000, "avg": 57.5, "matches": 270},
            },
        }

    async def get_team(self, team_id: str) -> Optional[dict]:
        """Get team details and statistics - mock implementation."""
        return {
            "id": team_id,
            "name": "India",
            "shortName": "IND",
            "country": "India",
            "logo": "https://via.placeholder.com/64",
            "stats": {
                "t20i": {"matches": 100, "wins": 65, "losses": 35},
                "odi": {"matches": 500, "wins": 280, "losses": 220},
            },
        }

    async def get_team_squad(self, team_id: str) -> List[dict]:
        """Get team squad for a series - mock implementation."""
        return [
            {
                "id": f"player_{i}",
                "name": f"Player {i}",
                "role": ["batsman", "bowler", "all-rounder"][i % 3],
                "jerseyNumber": i + 1,
            }
            for i in range(15)
        ]

    async def get_series(self, series_id: str) -> Optional[dict]:
        """Get series/tournament details - mock implementation."""
        return {
            "id": series_id,
            "name": "ICC T20 World Cup",
            "slug": "icc-t20-wc-2026",
            "format": "T20",
            "status": "upcoming",
            "season": "2026",
            "startDate": "2026-06-01",
            "endDate": "2026-07-15",
            "logo": "https://via.placeholder.com/200",
        }

    async def get_standings(self, series_id: str) -> List[dict]:
        """Get standings/table for a series - mock implementation."""
        teams = ["India", "Pakistan", "Australia", "England"]
        return [
            {
                "position": i + 1,
                "team": team,
                "matches": 5,
                "wins": 5 - i,
                "losses": i,
                "points": (5 - i) * 2,
            }
            for i, team in enumerate(teams)
        ]

    async def get_series_matches(self, series_id: str, status: str = "all") -> List[dict]:
        """Get matches for a series/tournament."""
        all_matches = await self.get_upcoming_matches(days=30)
        
        # Filter by series if multiple are available
        series_matches = [m for m in all_matches if m.get("league", {}).get("id") == series_id or m.get("series", {}).get("id") == series_id]
        
        if status != "all":
            series_matches = [m for m in series_matches if m.get("status") == status]
        
        return series_matches if series_matches else all_matches[:6]

    async def get_news(self, limit: int = 20) -> List[dict]:
        """Get cricket news - mock implementation."""
        return [
            {
                "id": f"news_{i}",
                "title": f"Cricket News {i}: Breaking Story",
                "description": "Exciting developments in international cricket",
                "source": "Cricbuzz",
                "imageUrl": "https://via.placeholder.com/200",
                "url": f"https://example.com/news/{i}",
                "publishedAt": (datetime.now() - timedelta(hours=i)).isoformat(),
            }
            for i in range(min(limit, 20))
        ]

    async def search(self, query: str) -> dict:
        """Search for matches, players, teams, news - mock implementation."""
        return {
            "matches": [{"id": "m1", "homeTeam": "India", "awayTeam": "Pakistan"}],
            "players": [{"id": "p1", "name": "Virat Kohli"}],
            "teams": [{"id": "t1", "name": "India"}],
            "news": [{"id": "n1", "title": "India wins match"}],
        }

    async def health_check(self) -> bool:
        """Check if provider is accessible - mock always returns True."""
        return True
